Mathlib/Combinatorics/Digraph/Basic.lean

```diff
@@ -0,0 +1,236 @@
+/-
+Copyright (c) 2024 Kyle Miller, Jack Cheverton. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kyle Miller, Jack Cheverton, Jeremy Tan
+-/
+import Mathlib.Order.CompleteBooleanAlgebra
+import Mathlib.Data.Fintype.Pi
+
+/-!
+# Digraphs
+
+This module defines directed graphs on a vertex type `V`. Although the structure is definitionally
+equivalent to a `Quiver` in a specialized universe, the intention is to consider graph theory
+language as basic. Furthermore, `Digraph` is not intended to be used as a class.
+
+## Main definitions
+
+* `Digraph` is a structure for relations. Unlike `SimpleGraph`, the relation does not need to be
+  symmetric or irreflexive.
+
+* `CompleteAtomicBooleanAlgebra` instance: Under the subgraph relation, `Digraph` forms a
+  `CompleteAtomicBooleanAlgebra`. In other words, this is the complete lattice of spanning subgraphs
+  of the complete graph.
+-/
+
+open Finset Function
+
+/-- A digraph is a relation `Adj` on a vertex type `V`.
+The relation describes which pairs of vertices are adjacent. -/
+@[ext]
+structure Digraph (V : Type*) where
+  /-- The adjacency relation of a digraph. -/
+  Adj : V → V → Prop
+
+noncomputable instance {V : Type*} [Fintype V] : Fintype (Digraph V) := by
+  classical exact Fintype.ofInjective Digraph.Adj (fun _ _ ↦ Digraph.ext)
+
+/-- Constructor for digraphs using a boolean function. -/
+@[simps]
+def Digraph.mk' {V : Type*} : (V → V → Bool) ↪ Digraph V where
+  toFun x := ⟨fun v w ↦ x v w⟩
+  inj' := by
+    intro adj adj'
+    simp only [mk.injEq]
+    intro h
+    funext v w
+    simpa only [eq_iff_iff, Bool.coe_iff_coe] using congr_fun₂ h v w
+
+/-- The complete digraph on a type `V` is the digraph with all pairs of distinct vertices
+adjacent. In `Mathlib`, this is usually referred to as `⊤`. -/
+def completeDigraph (V : Type*) : Digraph V where Adj := ⊤
+
+/-- The digraph with no edges on a given vertex type `V`. `Mathlib` prefers the notation `⊥`. -/
+def emptyDigraph (V : Type*) : Digraph V where Adj _ _ := False
+
+/-- Two vertices are adjacent in the complete bipartite digraph on two vertex types
+if and only if they are not from the same side.
+Any bipartite digraph may be regarded as a subgraph of one of these. -/
+@[simps]
+def Digraph.completeBipartiteGraph (V W : Type*) : Digraph (Sum V W) where
+  Adj v w := v.isLeft ∧ w.isRight ∨ v.isRight ∧ w.isLeft
+
+namespace Digraph
+
+variable {ι : Sort*} {V W X : Type*} (G : Digraph V) (G' : Digraph W) {a b c u v w : V}
+
+theorem adj_injective : Injective (Adj : Digraph V → V → V → Prop) :=
+  fun _ _ ↦ Digraph.ext
+
+@[simp]
+theorem adj_inj {G H : Digraph V} : G.Adj = H.Adj ↔ G = H :=
+  adj_injective.eq_iff
+
+section Order
+
+/-- The relation that one `Digraph` is a subgraph of another.
```

Comment:
Perhaps the docstring should mention that this is a spanning subgraph, not any subgraph